compiler/executable/match_/planner/plan.rs

```diff
@@ -577,48 +603,74 @@ impl<'a> ConjunctionPlanBuilder<'a> {
+                                }
+                            }
+                        }
                     } else {
-                        plan.clone_and_extend_with_new_step(extension, &self.graph)
-                    };
-
-                    if new_plans_heap.len() < beam_width {
-                        new_plans_heap.push(new_plan);
-                    } else if let Some(top) = new_plans_heap.peek() {
-                        if new_plan < *top {
-                            new_plans_heap.pop();
-                            new_plans_heap.push(new_plan);
+                        for extension in extension_heap.drain() {
+                            let new_plan = if !extension.is_constraint(&self.graph) {
+                                plan.clone_and_extend_with_new_step(extension, &self.graph)
+                            } else if extension.step_join_var.is_some()
+                                && (plan.ongoing_step_join_var.is_none()
+                                    || plan.ongoing_step_join_var == extension.step_join_var)
+                            {
+                                plan.clone_and_extend_with_continued_step(extension, &self.graph)
+                            } else {
+                                plan.clone_and_extend_with_new_step(extension, &self.graph)
+                            };
+
+                            let new_plan_hash = new_plan.hash();
+                            if !new_plans_hashset.contains(&new_plan_hash) {
+                                new_plans_hashset.insert(new_plan_hash);
+                                if new_plans_heap.len() < beam_width {
+                                    new_plans_heap.push(new_plan);
+                                } else if let Some(top) = new_plans_heap.peek() {
+                                    if new_plan < *top {
+                                        new_plans_heap.pop();
+                                        new_plans_heap.push(new_plan);
+                                    }
+                                }
+                            }
                         }
                     }
+                } else {
+                    unreachable!("Pattern extensions can never be empty, so a minimum should exist")
                 }
             }
             best_partial_plans = new_plans_heap.into_vec();
         }
 
         let best_plan = best_partial_plans.into_iter().min().unwrap();
-        let complete_plan = best_plan.into_complete_plan();
+        let complete_plan = best_plan.into_complete_plan(&self.graph);
         (complete_plan.vertex_ordering, complete_plan.pattern_metadata)
     }
 
     // Execute plans
     pub(super) fn plan(self) -> ConjunctionPlan<'a> {
         // Beam plan
+        let duration = Instant::now();
         let (ordering, metadata) = self.beam_search_plan();
+        println!("Planning took {:?}", duration.elapsed());
```

Comment:
some random printlns that if you want to keep should use event! with a LEVEL and we can enable/disable with the logging configuration